File: sdks/python/tikeo/src/tikeo/management.py
"""Management API helpers for SDK-side app-scoped jobs."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any
from urllib.parse import quote

import requests

from .errors import ManagementRequestError

API_KEY_HEADER = "x-tikeo-api-key"


@dataclass(slots=True)
class JobRetryPolicy:
    enabled: bool = True
    max_attempts: int = 3
    initial_delay_seconds: int = 5
    backoff_multiplier: int = 2
    max_delay_seconds: int = 60

    def to_json(self) -> dict[str, Any]:
        return {
            "enabled": self.enabled,
            "maxAttempts": self.max_attempts,
            "initialDelaySeconds": self.initial_delay_seconds,
            "backoffMultiplier": self.backoff_multiplier,
            "maxDelaySeconds": self.max_delay_seconds,
        }


def default_job_retry_policy() -> JobRetryPolicy:
    return JobRetryPolicy()


@dataclass(slots=True)
class JobDefinition:
    id: str = ""
    namespace: str = ""
    app: str = ""
    name: str = ""
    schedule_type: str = ""
    schedule_expr: str | None = None
    processor_name: str | None = None
    processor_type: str | None = None
    script_id: str | None = None
    enabled: bool = True
    retry_policy: JobRetryPolicy | None = None


@dataclass(slots=True)
class JobInstance:
    id: str = ""
    job_id: str = ""
    status: str = ""
    trigger_type: str = ""
    execution_mode: str = ""
    created_at: str = ""
    updated_at: str = ""


@dataclass(slots=True)
class CreateJobRequest:
    name: str
    schedule_type: str = "api"
    schedule_expr: str | None = None
    processor_name: str | None = None
    processor_type: str | None = None
    script_id: str | None = None
    enabled: bool = True
    retry_policy: JobRetryPolicy | None = None


@dataclass(slots=True)
class BroadcastSelectorRequest:
    tags: list[str] | None = None
    region: str | None = None
    cluster: str | None = None
    labels: dict[str, str] | None = None

    def to_json(self) -> dict[str, Any]:
        payload = {
            "tags": self.tags,
            "region": self.region,
            "cluster": self.cluster,
            "labels": self.labels,
        }
        return {key: value for key, value in payload.items() if value is not None}


@dataclass(slots=True)
class TriggerJobRequest:
    trigger_type: str = "api"
    execution_mode: str = "single"
    broadcast_selector: BroadcastSelectorRequest | None = None

    def to_json(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "triggerType": self.trigger_type,
            "executionMode": self.execution_mode,
        }
        if self.broadcast_selector is not None:
            payload["broadcastSelector"] = self.broadcast_selector.to_json()
        return payload


def api_trigger() -> TriggerJobRequest:
    return TriggerJobRequest()


def broadcast_api_trigger(selector: BroadcastSelectorRequest | None = None) -> TriggerJobRequest:
    return TriggerJobRequest(execution_mode="broadcast", broadcast_selector=selector)


def api_job(name: str, processor_name: str) -> CreateJobRequest:
    return CreateJobRequest(name=name, processor_name=processor_name, retry_policy=default_job_retry_policy())


def plugin_api_job(name: str, processor_type: str, processor_name: str) -> CreateJobRequest:
    return CreateJobRequest(name=name, processor_type=processor_type, processor_name=processor_name, retry_policy=default_job_retry_policy())


def script_api_job(name: str, script_id: str) -> CreateJobRequest:
    return CreateJobRequest(name=name, script_id=script_id, retry_policy=default_job_retry_policy())


class ManagementClient:
    def __init__(self, endpoint: str, api_key: str, namespace: str = "default", app: str = "default") -> None:
        self.endpoint = endpoint.strip().rstrip("/")
        self.api_key = api_key
        self.namespace = namespace.strip() or "default"
        self.app = app.strip() or "default"
        self.session = requests.Session()
        self.session.headers.update({"accept": "application/json", API_KEY_HEADER: api_key})

    def list_jobs(self) -> list[JobDefinition]:
        data = self._send("GET", "/jobs")
        items = data.get("items", []) if isinstance(data, dict) else []
        return [self._parse_job(item) for item in items if item.get("namespace") == self.namespace and item.get("app") == self.app]

    def create_job(self, request: CreateJobRequest) -> JobDefinition:
        payload = {
            "namespace": self.namespace,
            "app": self.app,
            "name": request.name,
            "scheduleType": request.schedule_type,
            "scheduleExpr": request.schedule_expr,
            "processorName": request.processor_name,
            "processorType": request.processor_type,
            "scriptId": request.script_id,
            "enabled": request.enabled,
            "retryPolicy": request.retry_policy.to_json() if request.retry_policy else None,
        }
        payload = {key: value for key, value in payload.items() if value is not None}
        return self._parse_job(self._send("POST", "/jobs", payload))

    def trigger_job(self, job_id: str, request: TriggerJobRequest | None = None) -> JobInstance:
        body = (request or api_trigger()).to_json()
        return self._parse_instance(self._send("POST", f"/jobs/{quote(job_id, safe='')}:trigger", body))

    def _send(self, method: str, path: str, body: dict[str, Any] | None = None) -> Any:
        response = self.session.request(method, f"{self.endpoint}/api/v1{path}", json=body, timeout=30)
        try:
            envelope = response.json()
        except ValueError as exc:
            raise ManagementRequestError(f"tikeo management response was not JSON: {response.status_code}") from exc
        if response.status_code < 200 or response.status_code >= 300 or envelope.get("code") != 0:
            raise ManagementRequestError(f"tikeo management request failed: status={response.status_code} message={envelope.get('message', '')}")
        if envelope.get("data") is None:
            raise ManagementRequestError("tikeo management response data was null")
        return envelope["data"]

    @staticmethod
    def _parse_job(data: dict[str, Any]) -> JobDefinition:
        retry = data.get("retryPolicy")
        return JobDefinition(
            id=data.get("id", ""),
            namespace=data.get("namespace", ""),
            app=data.get("app", ""),
            name=data.get("name", ""),
            schedule_type=data.get("scheduleType", ""),
            schedule_expr=data.get("scheduleExpr"),
            processor_name=data.get("processorName"),
            processor_type=data.get("processorType"),
            script_id=data.get("scriptId"),
            enabled=bool(data.get("enabled", True)),
            retry_policy=JobRetryPolicy(
                enabled=retry.get("enabled", True),
                max_attempts=retry.get("maxAttempts", 3),
                initial_delay_seconds=retry.get("initialDelaySeconds", 5),
                backoff_multiplier=retry.get("backoffMultiplier", 2),
                max_delay_seconds=retry.get("maxDelaySeconds", 60),
            ) if isinstance(retry, dict) else None,
        )

    @staticmethod
    def _parse_instance(data: dict[str, Any]) -> JobInstance:
        return JobInstance(
            id=data.get("id", ""),
            job_id=data.get("jobId", ""),
            status=data.get("status", ""),
            trigger_type=data.get("triggerType", ""),
            execution_mode=data.get("executionMode", ""),
            created_at=data.get("createdAt", ""),
            updated_at=data.get("updatedAt", ""),
        )

File: sdks/rust/tikee/src/management.rs
//! HTTP management client for app-scoped SDK API-key access.

use std::time::Duration;

use reqwest::header::{ACCEPT, CONTENT_TYPE, HeaderMap, HeaderValue};
use serde::{Deserialize, Serialize, de::DeserializeOwned};

use crate::error::WorkerSdkError;

const API_KEY_HEADER: &str = "x-tikee-api-key";

/// App-scoped management client authenticated with `X-Tikee-API-Key`.
#[derive(Debug, Clone)]
pub struct ManagementClient {
    http: reqwest::Client,
    endpoint: String,
    api_key: String,
    namespace: String,
    app: String,
}

impl ManagementClient {
    /// Build a management client for one namespace/app API key scope.
    #[must_use]
    pub fn new(
        endpoint: impl Into<String>,
        api_key: impl Into<String>,
        namespace: impl Into<String>,
        app: impl Into<String>,
    ) -> Self {
        Self {
            http: reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap_or_else(|_| reqwest::Client::new()),
            endpoint: trim_trailing_slash(&endpoint.into()),
            api_key: api_key.into(),
            namespace: defaulted(&namespace.into(), "default"),
            app: defaulted(&app.into(), "default"),
        }
    }

    /// List jobs visible to this API key.
    ///
    /// # Errors
    ///
    /// Returns an SDK error for transport, server, or decoding failures.
    pub async fn list_jobs(&self) -> Result<Vec<JobDefinition>, WorkerSdkError> {
        let page: Page<JobDefinition> = self
            .send(reqwest::Method::GET, "/jobs", Option::<&()>::None)
            .await?;
        Ok(page
            .items
            .into_iter()
            .filter(|job| job.namespace == self.namespace && job.app == self.app)
            .collect())
    }

    /// Create one job in the configured namespace/app scope.
    ///
    /// # Errors
    ///
    /// Returns an SDK error for transport, server, or decoding failures.
    pub async fn create_job(
        &self,
        request: CreateJobRequest,
    ) -> Result<JobDefinition, WorkerSdkError> {
        let payload = ScopedCreateJobRequest {
            namespace: &self.namespace,
            app: &self.app,
            name: &request.name,
            schedule_type: request.schedule_type.as_deref(),
            schedule_expr: request.schedule_expr.as_deref(),
            processor_name: request.processor_name.as_deref(),
            processor_type: request.processor_type.as_deref(),
            script_id: request.script_id.as_deref(),
            enabled: request.enabled,
            retry_policy: request.retry_policy.as_ref(),
        };
        self.send(reqwest::Method::POST, "/jobs", Some(&payload))
            .await
    }

    async fn send<T, B>(
        &self,
        method: reqwest::Method,
        path: &str,
        body: Option<&B>,
    ) -> Result<T, WorkerSdkError>
    where
        T: DeserializeOwned,
        B: Serialize + Sync + ?Sized,
    {
        let url = format!("{}/api/v1{}", self.endpoint, path);
        let mut headers = HeaderMap::new();
        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
        headers.insert(
            API_KEY_HEADER,
            HeaderValue::from_str(&self.api_key).map_err(|error| {
                WorkerSdkError::ManagementRequestFailed(format!("invalid api key header: {error}"))
            })?,
        );
        let mut request = self.http.request(method, url).headers(headers);
        if let Some(body) = body {
            request = request.header(CONTENT_TYPE, "application/json").json(body);
        }
        let response = request.send().await.map_err(|error| {
            WorkerSdkError::ManagementRequestFailed(format!("request failed: {error}"))
        })?;
        let status = response.status();
        let envelope: ApiEnvelope<T> = response.json().await.map_err(|error| {
            WorkerSdkError::ManagementRequestFailed(format!("response decode failed: {error}"))
        })?;
        if !status.is_success() || envelope.code != 0 {
            return Err(WorkerSdkError::ManagementRequestFailed(format!(
                "management request failed: status={status} message={}",
                envelope.message
            )));
        }
        envelope.data.ok_or_else(|| {
            WorkerSdkError::ManagementRequestFailed("management response data was null".to_owned())
        })
    }
}

/// Structured failure retry policy. `max_attempts` includes the first execution.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JobRetryPolicy {
    /// Whether failed executions should be retried.
    pub enabled: bool,
    /// Total attempts including the first execution.
    pub max_attempts: i32,
    /// Delay before the first retry attempt, in seconds.
    pub initial_delay_seconds: i64,
    /// Integer multiplier applied to each subsequent retry delay.
    pub backoff_multiplier: i32,
    /// Upper bound for any retry delay, in seconds.
    pub max_delay_seconds: i64,
}

impl Default for JobRetryPolicy {
    fn default() -> Self {
        Self {
            enabled: true,
            max_attempts: 3,
            initial_delay_seconds: 5,
            backoff_multiplier: 2,
            max_delay_seconds: 60,
        }
    }
}

/// Job definition returned by the tikee management API.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JobDefinition {
    /// Job id.
    pub id: String,
    /// Namespace name.
    pub namespace: String,
    /// App name.
    pub app: String,
    /// Display name.
    pub name: String,
    /// Schedule type.
    pub schedule_type: String,
    /// Optional schedule expression.
    pub schedule_expr: Option<String>,
    /// Optional processor binding.
    pub processor_name: Option<String>,
    /// Optional structured plugin processor type.
    pub processor_type: Option<String>,
    /// Optional script binding.
    pub script_id: Option<String>,
    /// Whether this job is enabled.
    pub enabled: bool,
    /// Structured failure retry policy applied to this job.
    pub retry_policy: JobRetryPolicy,
}

/// Job creation request for Rust management SDK users.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CreateJobRequest {
    /// Display name.
    pub name: String,
    /// Schedule type. Defaults server-side when omitted.
    pub schedule_type: Option<String>,
    /// Optional schedule expression.
    pub schedule_expr: Option<String>,
    /// Optional processor binding.
    pub processor_name: Option<String>,
    /// Optional structured plugin processor type.
    pub processor_type: Option<String>,
    /// Optional script binding.
    pub script_id: Option<String>,
    /// Optional enabled flag.
    pub enabled: Option<bool>,
    /// Optional structured failure retry policy.
    pub retry_policy: Option<JobRetryPolicy>,
}

impl CreateJobRequest {
    /// Build an API-triggered processor job request.
    #[must_use]
    pub fn api(name: impl Into<String>, processor_name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            schedule_type: Some("api".to_owned()),
            schedule_expr: None,
            processor_name: Some(processor_name.into()),
            processor_type: None,
            script_id: None,
            enabled: Some(true),
            retry_policy: Some(JobRetryPolicy::default()),
        }
    }

    /// Build an API-triggered plugin processor job request.
    #[must_use]
    pub fn plugin_api(
        name: impl Into<String>,
        processor_type: impl Into<String>,
        processor_name: impl Into<String>,
    ) -> Self {
        Self {
            name: name.into(),
            schedule_type: Some("api".to_owned()),
            schedule_expr: None,
            processor_name: Some(processor_name.into()),
            processor_type: Some(processor_type.into()),
            script_id: None,
            enabled: Some(true),
            retry_policy: Some(JobRetryPolicy::default()),
        }
    }

    /// Build an API-triggered script job request.
    #[must_use]
    pub fn script_api(name: impl Into<String>, script_id: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            schedule_type: Some("api".to_owned()),
            schedule_expr: None,
            processor_name: None,
            processor_type: None,
            script_id: Some(script_id.into()),
            enabled: Some(true),
            retry_policy: Some(JobRetryPolicy::default()),
        }
    }
}

#[derive(Debug, Deserialize)]
struct ApiEnvelope<T> {
    code: i32,
    message: String,
    data: Option<T>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct Page<T> {
    items: Vec<T>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ScopedCreateJobRequest<'a> {
    namespace: &'a str,
    app: &'a str,
    name: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    schedule_type: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    schedule_expr: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    processor_name: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    processor_type: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    script_id: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    retry_policy: Option<&'a JobRetryPolicy>,
}

fn trim_trailing_slash(value: &str) -> String {
    value.trim_end_matches('/').to_owned()
}

fn defaulted(value: &str, fallback: &str) -> String {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        fallback.to_owned()
    } else {
        trimmed.to_owned()
    }
}
